### PR TITLE
Strong Inv Pool Example

### PR DESCRIPTION
**Pull request by nivcertora**
https://certora.atlassian.net/browse/CERT-6867

### CODE REVIEW

**Review by yoav-el-certora — CHANGES_REQUESTED** [review]
We should keep the CVLByExample/StrongInvariants repo.
This is the link we published to clients in our latest release, and we should have an example for that when searching in Examples repo.

**Comment by nivcertora** [comment]
> We should keep the CVLByExample/StrongInvariants repo. This is the link we published to clients in our latest release, and we should have an example for that when searching in Examples repo.

This example is wrong...

**Comment by yoav-el-certora** [comment]
@nd-certora Can you please advise?

File: DEFI/LiquidityPool/README.md [hunk 1]
```diff
@@ -71,4 +71,82 @@ This example will verify the spec with a malicious receiver that transfers the m
...
+    assert Inv;  // <----- end block 1
+    require(IFlashLoanReceiver(receiverAddress).executeOperation(amount, totalPremium, msg.sender), 'P_INVALID_FLASH_LOAN_EXECUTOR_RETURN');  // <--- HAVOC_ECF
+    require Inv;  // <----- start block 2
+    asset.transferFrom(msg.sender, address(this), amountPlusPremium);
+    depositedAmount += amountPlusPremium;
+    assert Inv;  // <----- end block 2
+}
+```
+
+In summary, the strong invariant ensures that the invariant holds during both block 1 and block 2 of the function execution. It assumes that the external call cannot modify the contract storage and break the invariant.
```
Comment (nd-certora) [inline DEFI/LiquidityPool/README.md:114]:
```suggestion
In summary, the strong invariant ensures that the invariant holds during both block 1 and block 2 of the function execution. By induction, the Prover assumes that the external call may modify the contract storage but that the invariant holds afterward. 
```

File: DEFI/LiquidityPool/README.md [hunk 2]
```diff
@@ -71,4 +71,82 @@ This example will verify the spec with a malicious receiver that transfers the m
...
+```solidity
+function flashLoan(address receiverAddress, uint256 amount) nonReentrant() public {     
+    require Inv;  // <----- start block 1     
+    uint256 totalPremium = calcPremium(amount);
+    require(totalPremium != 0);
+    uint256 amountPlusPremium = amount + totalPremium;
+    asset.transferFrom(address(this), msg.sender, amount);
+    depositedAmount -= amount;
+    assert Inv;  // <----- end block 1
+    require(IFlashLoanReceiver(receiverAddress).executeOperation(amount, totalPremium, msg.sender), 'P_INVALID_FLASH_LOAN_EXECUTOR_RETURN');  // <--- HAVOC_ECF
```
Comment (nd-certora) [inline DEFI/LiquidityPool/README.md:106]: HAVOC_ALL (all contracts are havoced)

File: DEFI/LiquidityPool/README.md [hunk 3]
```diff
@@ -71,4 +71,82 @@ This example will verify the spec with a malicious receiver that transfers the m
...
+
+```solidity
+require Inv;  // <----- start block 1
+function flashLoan(address receiverAddress, uint256 amount) nonReentrant() public {          
+    uint256 totalPremium = calcPremium(amount);
+    require(totalPremium != 0);
+    uint256 amountPlusPremium = amount + totalPremium;
+    asset.transferFrom(address(this), msg.sender, amount);
+    depositedAmount -= amount;
+    require(IFlashLoanReceiver(receiverAddress).executeOperation(amount, totalPremium, msg.sender), 'P_INVALID_FLASH_LOAN_EXECUTOR_RETURN');  // <--- HAVOC_ALL
```
Comment (nd-certora) [inline DEFI/LiquidityPool/README.md:130]: here it is HAVOC_ECF (all contracts but the main one)

File: DEFI/LiquidityPool/README.md [hunk 4]
```diff
@@ -71,4 +71,82 @@ This example will verify the spec with a malicious receiver that transfers the m
...
+    asset.transferFrom(msg.sender, address(this), amountPlusPremium);
+    depositedAmount += amountPlusPremium;
+}
+assert Inv;  // <----- end block 1
+```
+
+In summary, the weak invariant ensures that the invariant holds before the function call and after it finishes, without making any assumptions about the external call.
+
+This version of the invariant does not hold during the verification process.
```
Comment (nd-certora) [inline DEFI/LiquidityPool/README.md:139]: lets explain why

File: DEFI/LiquidityPool/README.md (hunk 2 above)
Comment (nivcertora) [reply DEFI/LiquidityPool/README.md]: Done

File: DEFI/LiquidityPool/README.md (hunk 3 above)
Comment (nivcertora) [reply DEFI/LiquidityPool/README.md]: Done

File: DEFI/LiquidityPool/README.md (hunk 4 above)
Comment (nivcertora) [reply DEFI/LiquidityPool/README.md:139]: Done

File: DEFI/LiquidityPool/README.md [hunk 5]
```diff
@@ -71,4 +71,84 @@ This example will verify the spec with a malicious receiver that transfers the m
...
+    asset.transferFrom(address(this), msg.sender, amount);
+    depositedAmount -= amount;
+    require(IFlashLoanReceiver(receiverAddress).executeOperation(amount, totalPremium, msg.sender), 'P_INVALID_FLASH_LOAN_EXECUTOR_RETURN');  // <--- HAVOC_ECF
+    asset.transferFrom(msg.sender, address(this), amountPlusPremium);
+    depositedAmount += amountPlusPremium;
+}
+assert Inv;  // <----- end block 1
+```
+
+In summary, the weak invariant ensures that the invariant holds before the function call and after it finishes, without making any assumptions about the external call.
```
Comment (nd-certora) [inline DEFI/LiquidityPool/README.md:137]:
```suggestion
In general, the weak invariant ensures that the invariant holds after a function finishes where external call are modeled according to the definitions provided (or the Prover's defaults) 
```

File: DEFI/LiquidityPool/certora/specs/strong_inv.spec [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+methods
+{
+    function underlying.balanceOf(address)           external returns(uint256) envfree;
+    function depositedAmount()                       external returns(uint256) envfree;
+}
+
+strong invariant strongDepositedAmountLessThanContractUnderlyingAsset()
+    depositedAmount() <= underlying.balanceOf(currentContract)
+    filtered { f -> f.selector == sig:flashLoan(address,uint256).selector }
```
Comment (nd-certora) [inline DEFI/LiquidityPool/certora/specs/strong_inv.spec:11]: remove filter

File: DEFI/LiquidityPool/certora/specs/strong_inv.spec [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        preserved with(env e) {
+            require e.msg.sender != currentContract;
+        }
+    }
+
+weak invariant weakDepositedAmountLessThanContractUnderlyingAsset()
+    depositedAmount() <= underlying.balanceOf(currentContract)
+    filtered { f -> f.selector == sig:flashLoan(address,uint256).selector }
```
Comment (nd-certora) [inline DEFI/LiquidityPool/certora/specs/strong_inv.spec:20]: remover filter

File: DEFI/LiquidityPool/certora/specs/strong_inv.spec (hunk 6 above)
Comment (nivcertora) [reply DEFI/LiquidityPool/certora/specs/strong_inv.spec]: Done

File: DEFI/LiquidityPool/certora/specs/strong_inv.spec (hunk 7 above)
Comment (nivcertora) [reply DEFI/LiquidityPool/certora/specs/strong_inv.spec]: Done

**Review by nd-certora — APPROVED** [review]

**Pull request merged**